🐛 Bug: Incorrect information from `package-json/no-empty-fields` for `"browserslist": []`

### Bug Report Checklist

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have pulled the latest `main` branch of the repository.
- [x] I have [searched for related issues](https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues?q=is%3Aissue) and found none that matched my issue.

### Expected

`package-json/no-empty-fields` is correct in saying that having `"browserslist": []` in package.json does nothing and can be removed.

### Actual

Removing `"browserslist": []` from package.json results in different behavior, equivalent to having `"browserslist": [ "defaults" ]` in there.

### Additional Info

### Reproduction

1. Create package.json with this content:
    ```json
    {
        "browserslist": []
    }
    ```
2. Create eslint.config.mjs with this content:
    ```js
    import packageJson from "eslint-plugin-package-json";

    export default [
            packageJson.configs.recommended,
    ];
    ```
3. Run `npm add browserslist eslint eslint-plugin-package-json`.
4. Run `npx browserslist`. Observe that there is no output from the command.
5. Run `npx eslint`. Observe that the output includes
   ```
     2:5  error  The field 'browserslist' does nothing and can be removed  package-json/no-empty-fields
   ```
6. As suggested, remove the field from package.json.
7. Run `npx browserslist` again. Observe that the command lists various browsers now.

You can also reproduce with something like
    ```json
        "browserslist": {
            "development": [],
            "production": [ "last 1 version", "> 1%", "not dead" ],
        }
    ```
in step 1 and running `BROWSERSLIST_ENV=development npx browserslist` in steps 4 and 7.

### Additional notes

It may be sufficient for the rule to have a configurable list of fields to ignore, rather than trying to have the rule know about every possible field that might treat an empty array or empty object differently from the field being unset.

--- a/docs/rules/no-empty-fields.md
+++ b/docs/rules/no-empty-fields.md
@@ -33,3 +33,22 @@ Example of **correct** code for this rule:
 	}
 }
 ```
+
+## Options
+
+### `ignoreProperties`
+
+Pass an array of top-level package properties to ignore.
+When provided, any errors related to the properties won't cause the rule to report a violation.
+This can be useful if you're using tools that take configuration from package.json and accept an empty array or object as valid non-default configuration.
+
+Example of excluding the browserslist property used by [browserslist](https://www.npmjs.com/package/browserslist):
+
+```json
+{
+	"package-json/no-empty-fields": [
+		"error",
+		{ "ignoreProperties": ["browserslist"] }
+	]
+}
+```

--- a/docs/rules/no-empty-fields.md
+++ b/docs/rules/no-empty-fields.md
@@ -39,7 +39,7 @@ Example of **correct** code for this rule:
 ### `ignoreProperties`
 
 Pass an array of top-level package properties to ignore.
-When provided, any errors related to the properties won't cause the rule to report a violation.
+When provided, the rule won't report violations for the specified properties.
 This can be useful if you're using tools that take configuration from package.json and accept an empty array or object as valid non-default configuration.
 
 Example of excluding the browserslist property used by [browserslist](https://www.npmjs.com/package/browserslist):

--- a/src/rules/no-empty-fields.ts
+++ b/src/rules/no-empty-fields.ts
@@ -94,9 +94,7 @@ const getTopLevelPropertyName = (
 	) {
 		n = n.parent;
 	}
-	return n.type === "JSONProperty"
-		? (n.key as JsonAST.JSONStringLiteral).value
-		: undefined;
+	return ((n as JsonAST.JSONProperty).key as JsonAST.JSONStringLiteral).value;
 };
 
 export const rule = createRule<Options>({
@@ -107,10 +105,7 @@ export const rule = createRule<Options>({
 			JSONArrayExpression(node: JsonAST.JSONArrayExpression) {
 				if (!node.elements.length) {
 					const topLevelProperty = getTopLevelPropertyName(node);
-					if (
-						topLevelProperty === undefined ||
-						!ignoreProperties.includes(topLevelProperty)
-					) {
+					if (!ignoreProperties.includes(topLevelProperty)) {
 						report(context, getNode(node));
 					}
 				}
@@ -118,10 +113,7 @@ export const rule = createRule<Options>({
 			JSONObjectExpression(node: JsonAST.JSONObjectExpression) {
 				if (!node.properties.length) {
 					const topLevelProperty = getTopLevelPropertyName(node);
-					if (
-						topLevelProperty === undefined ||
-						!ignoreProperties.includes(topLevelProperty)
-					) {
+					if (!ignoreProperties.includes(topLevelProperty)) {
 						report(context, getNode(node));
 					}
 				}

--- a/src/rules/no-empty-fields.ts
+++ b/src/rules/no-empty-fields.ts
@@ -8,6 +8,11 @@ import * as ESTree from "estree";
 
 import { createRule, PackageJsonRuleContext } from "../createRule.js";
 
+interface Option {
+	ignoreProperties?: string[];
+}
+type Options = [Option?];
+
 const getDataAndMessageId = (
 	node:
 		| JsonAST.JSONArrayExpression
@@ -79,17 +84,46 @@ const getNode = (
 	return node.parent.type === "JSONProperty" ? node.parent : node;
 };
 
-export const rule = createRule({
+const getTopLevelPropertyName = (
+	node: JsonAST.JSONArrayExpression | JsonAST.JSONObjectExpression,
+) => {
+	let n: JsonAST.JSONNode = node;
+	while (
+		n.parent.parent?.parent?.type !== undefined &&
+		n.parent.parent.parent.type !== "Program"
+	) {
+		n = n.parent;
+	}
+	return n.type === "JSONProperty"
+		? (n.key as JsonAST.JSONStringLiteral).value
+		: undefined;
+};
+
+export const rule = createRule<Options>({
 	create(context) {
+		const ignoreProperties = context.options[0]?.ignoreProperties ?? [];
+
 		return {
 			JSONArrayExpression(node: JsonAST.JSONArrayExpression) {
 				if (!node.elements.length) {
-					report(context, getNode(node));
+					const topLevelProperty = getTopLevelPropertyName(node);
+					if (
+						topLevelProperty === undefined ||
+						!ignoreProperties.includes(topLevelProperty)
+					) {
+						report(context, getNode(node));
+					}
 				}
 			},
 			JSONObjectExpression(node: JsonAST.JSONObjectExpression) {
 				if (!node.properties.length) {
-					report(context, getNode(node));
+					const topLevelProperty = getTopLevelPropertyName(node);
+					if (
+						topLevelProperty === undefined ||
+						!ignoreProperties.includes(topLevelProperty)
+					) {
+						report(context, getNode(node));
+					}
 				}
 			},
 		};
@@ -108,7 +142,20 @@ export const rule = createRule({
 				"The field '{{field}}' does nothing and can be removed.",
 			remove: "Remove this empty field.",
 		},
-		schema: [],
+		schema: [
+			{
+				additionalProperties: false,
+				properties: {
+					ignoreProperties: {
+						items: {
+							type: "string",
+						},
+						type: "array",
+					},
+				},
+				type: "object",
+			},
+		],
 		type: "suggestion",
 	},
 });

--- a/src/tests/rules/no-empty-fields.test.ts
+++ b/src/tests/rules/no-empty-fields.test.ts
@@ -293,6 +293,85 @@ ruleTester.run("no-empty-fields", rule, {
 				},
 			],
 		},
+		{
+			code: `{
+\t\t"name": "test",
+\t\t"files": [],
+\t\t"browserslist": [],
+\t\t"scripts": {}
+}
+`,
+			errors: [
+				{
+					messageId: "emptyFields",
+					suggestions: [
+						{
+							messageId: "remove",
+							output: `{
+\t\t"name": "test",
+\t\t
+\t\t"browserslist": [],
+\t\t"scripts": {}
+}
+`,
+						},
+					],
+				},
+				{
+					messageId: "emptyFields",
+					suggestions: [
+						{
+							messageId: "remove",
+							output: `{
+\t\t"name": "test",
+\t\t"files": [],
+\t\t"browserslist": []
+\t\t
+}
+`,
+						},
+					],
+				},
+			],
+			options: [
+				{
+					ignoreProperties: ["browserslist"],
+				},
+			],
+		},
+		{
+			code: `{
+\t\t"name": "test",
+\t\t"browserslist": {
+\t\t\t"development": [],
+\t\t\t"production": [ "last 1 version", "> 1%", "not dead" ]
+\t\t}
+}
+`,
+			errors: [
+				{
+					messageId: "emptyFields",
+					suggestions: [
+						{
+							messageId: "remove",
+							output: `{
+\t\t"name": "test",
+\t\t"browserslist": {
+\t\t\t
+\t\t\t"production": [ "last 1 version", "> 1%", "not dead" ]
+\t\t}
+}
+`,
+						},
+					],
+				},
+			],
+			options: [
+				{
+					ignoreProperties: ["development"],
+				},
+			],
+		},
 	],
 	valid: [
 		`{ "name": "test", "files": ["index.js"] }`,
@@ -303,5 +382,21 @@ ruleTester.run("no-empty-fields", rule, {
 		`{ "name": "test", "peerDependencyMeta": { "@altano/repository-tools": { "optional": true } } }`,
 		`{ "name": "test", "peerDependencyMeta": { "@altano/repository-tools": { "optional": true, "test": [{"test": ["1"]}] } } }`,
 		`{ "name": "test", "peerDependencyMeta": { "@altano/repository-tools": { "optional": true, "test": ["field1"] } } }`,
+		{
+			code: `{ "name": "test", "browserslist": [] }`,
+			options: [
+				{
+					ignoreProperties: ["browserslist"],
+				},
+			],
+		},
+		{
+			code: `{ "name": "test", "browserslist": { "development": [], "production": [ "last 1 version", "> 1%", "not dead" ] } }`,
+			options: [
+				{
+					ignoreProperties: ["browserslist"],
+				},
+			],
+		},
 	],
 });

